### PR TITLE
[DeepSeekV3.2] Enable pure TP & Partial DP Attention

### PR DESCRIPTION
**Pull request by YAMY1234**
## Motivation

DeepSeekV3.2 NSA currently has rough edges when running in **pure TP mode** (`dp_size < tp_size`):

- FlashMLA sparse can see an invalid `num_heads` per rank after TP sharding.
- NSA's get_mla_metadata crashes with TP + large bs (Fixed with [Fix FlashMLA Shared-Memory Overflow in SGLang's Pure-TP Mode with Low-SMEM Fallback Scheduler #2](https://github.com/sgl-project/FlashMLA/pull/2))
- The NSA fp8 MQA indexer may OOM when building a full `(num_q, num_k)` logits matrix in large pure-TP batches.

This PR makes **NSA + pure TP & partial DP Attention** a supported and stable configuration for DeepSeekV3.2.
Should merge after [Upgrade flashmla kernel for NSA tp support #13718](https://github.com/sgl-project/sglang/pull/13718).

Sample launch commands:
```
python -m sglang.launch_server  --model deepseek-ai/DeepSeek-V3.2-Exp  --tp 8
```
```
python -m sglang.launch_server  --model deepseek-ai/DeepSeek-V3.2-Exp  --tp 8 --dp 4 --enable-dp-attention
```

---

## Modifications

- **NSA backend (`nsa_backend.py`)**
  - In `_forward_flashmla_sparse(...)`, pad `q`’s head dimension to the required multiple (64 on SM90, 128 on SM100+), call the sparse kernel with padded heads, then trim the output back to the original `num_heads`(TP support).
  - Extend `topk_transform(...)` with `topk_indices_offset_override` to accept precomputed ragged offsets(indexer chunking support).
  - Cache `device_capability` / `device_sm_major` on init and reuse it in TRTLLM ragged and FlashMLA paths.

- **Server args (`server_args.py`)**
  - Allow `dp_size < tp_size` for DeepSeekV3.2 NSA:
    - Log a warning for NSA + TP configuration.

- **FlashMLA cmake (`flashmla.cmake`)**
  - Bump FlashMLA submodule to `be055fb7df0090fde45f08e9cb5b8b4c0272da73` to use the latest sparse kernel(Avoid crashing with large bs).

- **NSA indexer (`nsa_indexer.py`)**
  - Add `_should_chunk_mqa_logits(...)` to decide when to chunk fp8 MQA logits based on workload size and free GPU memory.
  - In `_get_topk_ragged(...)`, add a chunked path that:
    - Computes fp8 MQA logits in slices to avoid OOM.
    - Uses `topk_indices_offset_override` so each chunk can reuse the global ragged offsets safely.
    - Writes into a preallocated `(token_nums, index_topk)` buffer for all tokens.
- Fix a shape-mismatch bug in NSA sparse prefill:  
  when MLP-sync pads tokens to TP multiples (`7→8`), the indexer still returns only real-token rows.  
  `_pad_topk_indices(...)`allow padding to maintain `#tokens(q) == #tokens(topk_indices)`, restoring correctness for FlashMLA-sparse under partial DP attention (e.g., TP 8, DP 4).

## Accuracy Tests

Launch with
```
python -m sglang.launch_server  --model deepseek-ai/DeepSeek-V3.2-Exp  --tp 8
```
```
GPQA:
Repeat: 8, mean: 0.799                                                                                                            | 26/198 [36:38<1:39:53, 34.84s/it]
Scores: ['0.778', '0.798', '0.823', '0.803', '0.778', '0.803', '0.788', '0.823']
====================█▎                                                                                                           | 26/198 [37:46<6:15:40, 131.05s/it]
Writing report to /tmp/gpqa_deepseek-ai_DeepSeek-V3.2-Exp.html                                                                    | 48/198 [37:35<2:00:18, 48.12s/it]
{'chars': np.float64(14523.828282828283), 'chars:std': np.float64(11247.561643307768), 'score:std': np.float64(0.38147197173296354), 'score': np.float64(0.8232323232323232)}
Writing results to /tmp/gpqa_deepseek-ai_DeepSeek-V3.2-Exp.json
Total latency: 2267.919 s
Score: 0.823
```

## Benchmarking and Profiling

Serving with one request:
```
python3 -m sglang.bench_serving     --backend sglang     --dataset-name random     --num-prompts 1     --random-input-len 120000     --random-output-len 1     --disable-ignore-eos --random-range-ratio 1.0
```
TP mode:
```
python -m sglang.launch_server  --model deepseek-ai/DeepSeek-V3.2-Exp  --tp 8

Input token throughput (tok/s):          284607.82 
```
DP Attention:
```
python -m sglang.launch_server  --model deepseek-ai/DeepSeek-V3.2-Exp  --tp 8 --dp 8 --enable-dp-attention

Input token throughput (tok/s):          6609.56   
```

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.ai/developer_guide/contribution_guide.html#code-style-guidance).
- [ ] Work with maintainers to merge your PR. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)

### CODE REVIEW

**Comment by Fridge003** [comment]
Thanks @YAMY1234~ If your PR is blocked on FlashMLA side, you can create a new branch at https://github.com/sgl-project/FlashMLA. The flashmla kernel now integrated in sglang are built on this repo

**Comment by YAMY1234** [comment]
> > In _forward_flashmla_sparse(...), pad q’s head dimension to the required multiple (64 on SM90, 128 on SM100+)
> 
> @YAMY1234 For the Hxx device, the padding head may exhibit poor performance. Could we consider swapping the head and token dimensions via all-to-all (a2a) communication, where each rank increases the head dimension while reducing the token dimension?

Thanks for the suggestion!

For now, this padding logic only applies to the FlashMLA sparse path under pure TP / partial DP attention. For original normal DP attention path is unchanged — if the layout doesn’t match, it will just fail as before, so there should be no behavioral change there.

I agree that doing an a2a-based head/token swap could be a good follow-up optimization for the TP / partial DP case on Hxx to reduce the padding overhead. For this PR, I think we can keep the change scoped to making pure TP & partial DP attention functional and stable, and we can explore the a2a approach in a separate perf-focused change.

**Review by Fridge003 — APPROVED** [review]

**Comment by Fridge003** [comment]
> > @YAMY1234 Can you add a benchmark for bs=1? Expectedly pure TP should be faster than DP+TP
> 
> @Fridge003 Added in the PR description~

Oh I mean performance benchmark. You can test with `python3 -m sglang.test.send_one`

**Comment by YAMY1234** [comment]
> > > @YAMY1234 Can you add a benchmark for bs=1? Expectedly pure TP should be faster than DP+TP
> > 
> > 
> > @Fridge003 Added in the PR description~
> 
> Oh I mean performance benchmark. You can test with `python3 -m sglang.test.send_one`

Sorry my previous added benchmark disappeared😂 might be an saving error. I tested with sglang.bench_serving and looks like TP will be 4-5x faster than DP Attention during prefill in bs=1's situation. Could you take a second look at the PR desc? Thanks!

**Comment by Fridge003** [comment]
@YAMY1234 Thanks~
Since this PR will break the usage of deepseek v32, can you please change all the related usage (appending --dp argument) in the test cases (files with prefix `test_deepseek_v32`). 
A pure tp test can be added under `test/nightly` folder. The document also needs update

**Comment by YAMY1234** [comment]
@Fridge003  Thanks! Added docs and unittest~
